### PR TITLE
ENH: Index[bool]

### LINKED ISSUE #15890: API: support a BooleanIndex
xref #15888 
xref #15751 

We generally *don't* actually keep these around very much (instead we use a numpy boolean array), but would be nice for compat. (e.g. #15589; were we could return an actual Index rather than a boolean array).

```
In [31]: Index([True, False])
Out[31]: Index([True, False], dtype='object')
```

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [x] closes #15890
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit) for how to run them
- [x] whatsnew entry

### CODE REVIEW

**Comment by jreback** [comment]
> Thoughts on deprecation path?

i think it might be ok to actually just change this, are you finding that you need to change test behavior?

**Comment by jbrockmendel** [comment]
>  are you finding that you need to change test behavior?

Yes.  Largely around factorize and value_counts where this branch is currently very ad-hoc.

**Comment by jreback** [comment]
looks ok for 1.4?

**Comment by jbrockmendel** [comment]
> looks ok for 1.4?

I'd rather not rush it.

File: pandas/tests/series/methods/test_drop.py [hunk 1]
```diff
@@ -54,7 +54,9 @@ def test_drop_with_ignore_errors():
 
     # GH 8522
     ser = Series([2, 3], index=[True, False])
-    assert ser.index.is_object()
+    ser = Series([2, 3], index=[True, False])
```
Comment (jreback) [inline pandas/tests/series/methods/test_drop.py:57]: duplicate line here

**Review by jreback — APPROVED** [review]
minor comment. ready to go?

**Comment by jbrockmendel** [comment]
> minor comment. ready to go?

just pushed with a fix.  i think ready, yes.

**Comment by jreback** [comment]
can you rebase again (as the scipy fix merged)

**Review by mroeschke — APPROVED** [review]

**Comment by jreback** [comment]
thanks @jbrockmendel very nice!

**Pull request merged**